workflow/src/main/java/bio/terra/tanagra/workflow/BuildPathsForHierarchy.java

```diff
@@ -0,0 +1,325 @@
+    String hierarchyQuery = Files.readString(Path.of(options.getHierarchyQuery()));
+
+    Pipeline pipeline = Pipeline.create(options);
+
+    // read in all (parent, child) rows
+    PCollection<TableRow> parentChildBQrows =
+        pipeline.apply(
+            "read all (parent,child) rows",
+            BigQueryIO.readTableRows()
+                .fromQuery(hierarchyQuery)
+                .withMethod(BigQueryIO.TypedRead.Method.EXPORT)
+                .usingStandardSql());
+
+    // build a collection of KV<child,parent>
+    PCollection<KV<Long, String>> childParentKVs =
+        parentChildBQrows.apply(
+            "build (child,parent) KV pairs",
+            MapElements.into(
+                    TypeDescriptors.kvs(TypeDescriptors.longs(), TypeDescriptors.strings()))
+                .via(BuildPathsForHierarchy::bqRowToChildParentKV));
+
+    // read in all (node) rows
+    PCollection<TableRow> nodeBqRows =
+        pipeline.apply(
+            "read all (node) rows",
+            BigQueryIO.readTableRows()
+                .fromQuery(allNodesQuery)
+                .withMethod(BigQueryIO.TypedRead.Method.EXPORT)
+                .usingStandardSql());
+
+    // build a collection of KV<node,path> where path=node
+    PCollection<KV<Long, String>> nextNodePathKVs =
+        nodeBqRows.apply(
+            "build (node,path) KV pairs",
+            MapElements.into(
+                    TypeDescriptors.kvs(TypeDescriptors.longs(), TypeDescriptors.strings()))
+                .via(BuildPathsForHierarchy::bqRowToNodePathKV));
+
+    // define the CoGroupByKey tags
+    final TupleTag<String> pathTag = new TupleTag<>();
+    final TupleTag<String> parentTag = new TupleTag<>();
+
+    // iterate through each possible level of the hierarchy, adding up to one node to each path per
+    // iteration
+    for (int ctr = 0; ctr < options.getMaxHierarchyDepth(); ctr++) {
+      // do a CoGroupByKey join of the current node-path collection and the child-parent collection
+      PCollection<KV<Long, CoGbkResult>> nodeParentPathJoin =
+          KeyedPCollectionTuple.of(pathTag, nextNodePathKVs)
+              .and(parentTag, childParentKVs)
+              .apply(CoGroupByKey.create());
+
+      nextNodePathKVs =
+          nodeParentPathJoin.apply(
+              ParDo.of(
+                  new DoFn<KV<Long, CoGbkResult>, KV<Long, String>>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext context) {
+                      KV<Long, CoGbkResult> element = context.element();
+                      Long node = element.getKey();
+                      Iterator<String> pathTagIter = element.getValue().getAll(pathTag).iterator();
+                      Iterator<String> parentTagIter =
+                          element.getValue().getAll(parentTag).iterator();
+
+                      // there may be multiple possible next steps (i.e. the current node may have
```

Comment:
Some of the nodes in the standard hierarchy are polyhierarchical, e.g. conditions. Have you covered this with Brian/ Robert already? I think the currently designed model doesn't really handle this scenario - which maybe is reasonable..

Reply:
We haven't discussed it much so far, but I do know they exist. I think we do currently support this, in the sense that:
- All parent-child relationships are included in one of the hierarchy index tables, so if a node has >1 parent, it should show up in a `CHILD_OF` query for both parents.
- The node-path table does only specify one path for each node, but I think that's what we need for Tanagra because even if a node can show up in >1 place in a hierarchy tree, we need to highlight exactly one of those places when someone clicks the hierarchy icon.

I am not sure whether the ancestor-descendant hierarchy index table that is generated accounts for nodes with >1 parent.